cpmpy/transformations/linearize.py

```diff
@@ -5,190 +5,187 @@
-            lhs = Operator("wsum", [[1]*len(lhs.args), lhs.args])
-            cons_lhs = []
-        if _wsum_should(lhs):
-            lhs, cons_lhs = Operator("wsum",_wsum_make(lhs)), []
-        if lhs.name != "wsum":
-            lhs, cons_lhs = get_or_make_var(lhs)
-        # Rewrite of constraints:
-        #   c1 = lhs - rhs <= Mz - 1
-        #   c2 = lhs - rhs >= -M(1-z) + 1
-        c1 = Operator("wsum", [[-M, 1],       [z, 1]]) + lhs <= rhs      # TODO: this can be optimized, see https://github.com/CPMpy/cpmpy/issues/97
-        c2 = Operator("wsum", [[-M, (M - 1)], [z, 1]]) + lhs >= rhs      # TODO: this can be optimized, see https://github.com/CPMpy/cpmpy/issues/97
-
-        return linearize_constraint(flatten_constraint(cons_lhs)) + [c1, c2]
-
-    if cpm_expr.name in [">=", "<=", "=="] and \
-            isinstance(cpm_expr.args[0], Operator) and cpm_expr.args[0].name == "mul":
-        lhs, rhs = cpm_expr.args
-        if len(lhs.args) == 2:
-            # multiplication of var and constant
-            if _wsum_should(lhs):
-                return [Comparison(cpm_expr.name, Operator("wsum", _wsum_make(lhs)), rhs)]
-            return [cpm_expr]
+            # BV -> LinExpr
+            if isinstance(cond, _BoolVarImpl):
+                lin_sub = linearize_constraint(sub_expr, supported=supported, reified=True)
+                return [cond.implies(lin) for lin in lin_sub]
 
+    # comparisons
+    if isinstance(cpm_expr, Comparison):
         lhs, rhs = cpm_expr.args
-        var, constraints = get_or_make_var(lhs.args[0] * lhs.args[1])
-        # decompose long multiplication in several multiplications with two args
-        for arg in lhs.args[2:]:
-            var, cons = get_or_make_var(var * arg)
-            constraints += cons
 
-        return constraints + [Comparison(cpm_expr.name, var, rhs)]
+        if isinstance(lhs, _NumVarImpl) and isinstance(rhs, _NumVarImpl):
+            # bring numvar to lhs
+            lhs = 1 * lhs + -1 * rhs
+            rhs = 0
+
+        # linearize unsupported operators
+        elif isinstance(lhs, Operator) and lhs.name not in supported: # TODO: add mul, (abs?), (mod?), (pow?)
+
+            if lhs.name == "sub":
+                # convert to wsum
+                lhs = sum([1 * lhs.args[0] + -1 * lhs.args[1]])
+
+            elif lhs.name == "mul" and is_num(lhs.args[0]):
+                lhs = Operator("wsum",[[lhs.args[0]], [lhs.args[1]]])
+            else:
+                raise TransformationNotImplementedError(f"lhs of constraint {cpm_expr} cannot be linearized, should be any of {supported | set(['sub'])} but is {lhs}. Please report on github")
+
+        elif isinstance(lhs, GlobalConstraint) and lhs.name not in supported:
+            raise ValueError("Linearization of `lhs` not supported, run `cpmpy.transformations.decompose_global.decompose_global() first")
+
+        if is_num(lhs) or (isinstance(lhs, Operator) and lhs.name in {"sum","wsum"}):
```

Comment:
What if `lhs` is a numeric variable? Is this case covered by `is_num(lhs)`?

Reply:
This is covered above for now, but will adapt the code so its covered here. That way it is more uniform